File: pt6/1209.remove-all-adjacent-duplicates-in-string-ii.go
package main

import "strings"

func removeDuplicates(s string, k int) string {
	type run struct {
		ch    byte
		count int
	}

	stack := make([]run, 0, len(s))
	for i := 0; i < len(s); i++ {
		ch := s[i]
		if len(stack) > 0 && stack[len(stack)-1].ch == ch {
			stack[len(stack)-1].count++
		} else {
			stack = append(stack, run{ch: ch, count: 1})
		}

		if stack[len(stack)-1].count == k {
			stack = stack[:len(stack)-1]
		}
	}

	var builder strings.Builder
	builder.Grow(len(s))
	for _, item := range stack {
		for i := 0; i < item.count; i++ {
			builder.WriteByte(item.ch)
		}
	}
	return builder.String()
}

/*
Explanation

The stack stores compressed runs: character plus how many consecutive copies
survive so far. For each byte, either extend the top run or push a new run. If
the top run reaches k, pop it immediately.

This stack is ideal because deletions only affect the newest surviving run.
When a group is removed, the previous stack entry becomes adjacent to future
characters automatically, which handles chained deletions without rescanning.

Go detail: strings.Builder constructs the final string efficiently instead of
repeated string concatenation.

Edge cases: k == 1 removes everything; alternating characters never hit k;
after one pop, later characters can merge with the run revealed underneath.

Time complexity: O(n).
Space complexity: O(n) in the worst case for the stack.
*/

File: pt6/1209.remove-all-adjacent-duplicates-in-string-ii.java
import java.util.ArrayDeque;
import java.util.Deque;

class Solution {
    public String removeDuplicates(String s, int k) {
        Deque<int[]> stack = new ArrayDeque<>();

        for (char ch : s.toCharArray()) {
            if (!stack.isEmpty() && stack.peekLast()[0] == ch) {
                stack.peekLast()[1]++;
            } else {
                stack.addLast(new int[] {ch, 1});
            }

            if (stack.peekLast()[1] == k) {
                stack.removeLast();
            }
        }

        StringBuilder ans = new StringBuilder();
        for (int[] run : stack) {
            for (int i = 0; i < run[1]; i++) {
                ans.append((char) run[0]);
            }
        }
        return ans.toString();
    }
}

/*
Explanation

Use a stack of compressed runs: [character, count]. The next character either
extends the most recent run or starts a new run. When the count reaches k, pop
that run immediately.

The stack is ideal because deletions only affect the newest surviving run. If a
run is removed, the previous stack entry is automatically exposed and can merge
with future characters.

Java choices: ArrayDeque is the standard stack/deque implementation; a
StringBuilder avoids repeated string concatenation when reconstructing.

Edge cases: k == 1 removes all characters; alternating characters never form a
large run; chained deletions are handled by popping.

Time complexity: O(n).
Space complexity: O(n).
*/
